[Wasm-GC] Implement basic support in tables for GC types

takikawa opened the pull request:
#### dda771352e63f963e07d78df75adfbc1826598c8
<pre>
[Wasm-GC] Implement basic support in tables for GC types
<a href="https://bugs.webkit.org/show_bug.cgi?id=251041">https://bugs.webkit.org/show_bug.cgi?id=251041</a>

Reviewed by Justin Michaud.

Updates Wasm table type parsing so that defaultable ref types other than
externref and funcref are allowed. Also stores the Wasm element type in
TableInformation, as the element type cannot be deduced from other
information when other ref types are allowed.

The internal implementations of the ref.cast and extern.internalize
operations are also moved so that they can be used from table prototype
methods.

This also implements the part of the JS API for tables, which specifies
that when JS tries to store a value into a table with a GC type, it is
internalized and then cast to the appropriate type.

The typed function reference proposal also extends tables to allow
non-defaultable ref types as long as a table initializer expression is
provided. This is not yet supported in this patch.

* JSTests/wasm/function-references/table.js: Added.
(module):
(async testTypedFuncrefTable):
* JSTests/wasm/gc-spec-harness.js:
* JSTests/wasm/gc-spec-tests/ref_cast.wast.js: Added.
* JSTests/wasm/gc/arrays.js:
* JSTests/wasm/gc/i31.js:
* JSTests/wasm/gc/structs.js:
* JSTests/wasm/js-api/table.js:
(assert.throws):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::internalizeExternref):
(JSC::Wasm::TableInformation::TableInformation):
(JSC::Wasm::TableInformation::wasmType const):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::refCast):
(JSC::Wasm::externInternalize):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseTableHelper):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::Table::Table):
(JSC::Wasm::Table::tryCreate):
(JSC::Wasm::ExternRefTable::ExternRefTable):
(JSC::Wasm::FuncRefTable::FuncRefTable):
(JSC::Wasm::FuncRefTable::createFixedSized):
(JSC::Wasm::Table::wasmType const): Deleted.
* Source/JavaScriptCore/wasm/WasmTable.h:
(JSC::Wasm::Table::wasmType const):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::castReference):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeExports):
* Source/JavaScriptCore/wasm/js/WebAssemblyTableConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/js/WebAssemblyTablePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/262459@main">https://commits.webkit.org/262459@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44d1a9969bbd4bb49a03fd1a2a0100cd6f5a0509

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1531 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1560 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1614 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2479 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/1679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1623 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1619 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1474 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1393 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1382 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2286 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1395 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1366 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1306 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1306 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1394 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1406 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2459 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1471 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1438 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1283 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1570 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1354 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1373 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/340 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/390 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1493 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1612 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/454 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

takikawa commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -884,6 +884,8 @@ class TypeInformation {
     static std::optional<RefPtr<const RTT>> tryGetCanonicalRTT(TypeIndex);
     static RefPtr<const RTT> getCanonicalRTT(TypeIndex);
 
+    static bool castReference(JSValue, bool, TypeIndex);
```
I wasn't sure if this is the best place to put this function. Initially I thought of putting it in `WasmInstance.cpp` like many other runtime operations, but it turns out this cast operation does not actually depend on any instance state (because type definitions are global, cross-instance state).

Anyway, would be happy to take suggestions if there's a better place to put it.

takikawa commented on Source/JavaScriptCore/wasm/WasmFormat.h:
```diff
@@ -354,6 +354,17 @@ inline bool isDefaultableType(StorageType type)
     return true;
 }
 
+inline JSValue internalizeExternref(JSValue value)
```
Ditto on not being sure if this is the best place for this function (see `castReference`).

justinmichaud reviewed (approved):
r=me. In general, as this effort wraps up, I would love to see some extra release asserts thrown around that wasm gc is enabled. I am a bit worried that one missing case in the parser might be able enough to pwn us.

takikawa committed:
[Wasm-GC] Implement basic support in tables for GC types
https://bugs.webkit.org/show_bug.cgi?id=251041

Reviewed by Justin Michaud.

Updates Wasm table type parsing so that defaultable ref types other than
externref and funcref are allowed. Also stores the Wasm element type in
TableInformation, as the element type cannot be deduced from other
information when other ref types are allowed.

The internal implementations of the ref.cast and extern.internalize
operations are also moved so that they can be used from table prototype
methods.

This also implements the part of the JS API for tables, which specifies
that when JS tries to store a value into a table with a GC type, it is
internalized and then cast to the appropriate type.

The typed function reference proposal also extends tables to allow
non-defaultable ref types as long as a table initializer expression is
provided. This is not yet supported in this patch.

* JSTests/wasm/function-references/table.js: Added.
(module):
(async testTypedFuncrefTable):
* JSTests/wasm/gc-spec-harness.js:
* JSTests/wasm/gc-spec-tests/ref_cast.wast.js: Added.
* JSTests/wasm/gc/arrays.js:
* JSTests/wasm/gc/i31.js:
* JSTests/wasm/gc/structs.js:
* JSTests/wasm/js-api/table.js:
(assert.throws):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::internalizeExternref):
(JSC::Wasm::TableInformation::TableInformation):
(JSC::Wasm::TableInformation::wasmType const):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::refCast):
(JSC::Wasm::externInternalize):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseTableHelper):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::Table::Table):
(JSC::Wasm::Table::tryCreate):
(JSC::Wasm::ExternRefTable::ExternRefTable):
(JSC::Wasm::FuncRefTable::FuncRefTable):
(JSC::Wasm::FuncRefTable::createFixedSized):
(JSC::Wasm::Table::wasmType const): Deleted.
* Source/JavaScriptCore/wasm/WasmTable.h:
(JSC::Wasm::Table::wasmType const):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::castReference):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeExports):
* Source/JavaScriptCore/wasm/js/WebAssemblyTableConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/js/WebAssemblyTablePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/262459@main
